src/connection_string.ts

```diff
@@ -785,26 +775,792 @@ function parseURI(uri: string): { srv: boolean; url: URL; hosts: string[] } {
+  const illegalCharacters = new RegExp(String.raw`[:/?#\[\]@]`, 'gi');
+  if (username?.match(illegalCharacters)) {
+    throw new MongoParseError(`Username contains unescaped characters ${username}`);
+  }
+  if (!username || !password) {
+    const uriWithoutProtocol = uri.replace(`${protocol}://`, '');
+    if (uriWithoutProtocol.startsWith('@') || uriWithoutProtocol.startsWith(':')) {
+      throw new MongoParseError('URI contained empty userinfo section');
+    }
+  }
+
+  if (password?.match(illegalCharacters)) {
+    throw new MongoParseError('Password contains unescaped characters');
+  }
+
+  let authString = '';
+  if (typeof username === 'string') authString += username;
+  if (typeof password === 'string') authString += `:${password}`;
+
+  const srv = protocol.includes('srv');
+  const hostList = hosts.split(',');
+  const url = new URL(`${protocol.toLowerCase()}://${authString}@dummyHostname${rest}`);
+
+  if (srv && hostList.length !== 1) {
+    throw new MongoParseError('mongodb+srv URI cannot have multiple service names');
+  }
+  if (srv && hostList[0].includes(':')) {
+    throw new MongoParseError('mongodb+srv URI cannot have port number');
+  }
+
   return {
-    srv: protocol.includes('srv'),
-    url: new URL(`${protocol.toLowerCase()}://${authString}@dummyHostname${rest}`),
+    srv,
+    url,
     hosts: hosts.split(',')
   };
 }
 
+const TRUTHS = new Set(['true', 't', '1', 'y', 'yes']);
+const FALSEHOODS = new Set(['false', 'f', '0', 'n', 'no', '-1']);
+function getBoolean(name: string, value: unknown): boolean {
+  if (typeof value === 'boolean') return value;
+  const valueString = String(value).toLowerCase();
+  if (TRUTHS.has(valueString)) return true;
+  if (FALSEHOODS.has(valueString)) return false;
+  throw new TypeError(`For ${name} Expected stringified boolean value, got: ${value}`);
+}
+
+function getInt(name: string, value: unknown): number {
+  if (typeof value === 'number') return Math.trunc(value);
+  const parsedValue = Number.parseInt(String(value), 10);
+  if (!Number.isNaN(parsedValue)) return parsedValue;
+  throw new TypeError(`Expected ${name} to be stringified int value, got: ${value}`);
+}
+
+function getUint(name: string, value: unknown): number {
+  const parsedValue = getInt(name, value);
+  if (parsedValue < 0) {
+    throw new TypeError(`${name} can only be a positive int value, got: ${value}`);
+  }
+  return parsedValue;
+}
+
+function toRecord(value: string): Record<string, any> {
+  const record = Object.create(null);
+  const keyValuePairs = value.split(',');
+  for (const keyValue of keyValuePairs) {
+    const [key, value] = keyValue.split(':');
+    record[key] = value;
+  }
+  return record;
+}
+
+const DEFAULT_PK_FACTORY = {
+  createPk(): ObjectId {
+    // We prefer not to rely on ObjectId having a createPk method
+    return new ObjectId();
+  }
+};
+
+const defaultOptions = new Map<string, unknown>(
```

Comment:
You already have a registry of options below with the type, and an optional `transform`, why can't we put the `default` there rather than maintaining this list separately?

Reply:
Good idea, moved the defaults to the registry!